### PR TITLE
Worktree oauth integration

### PR DESCRIPTION
**Pull request by jmagar**
## Summary by cubic
Adds OAuth via `lab-auth`, unifies auth across `/mcp` and `/api`, and enforces fail‑closed, scope‑based checks for MCP tools. Also merges in main’s observability layer and fixes a stdio startup await bug.

- **New Features**
  - Shared `AuthLayer` guards both `/mcp` and `/api` via `AuthPolicy` (`LoopbackDev` or `Mounted`). In bearer-only we use `Mounted { auth_state: None }` so the static token and scope checks are enforced; OAuth routes mount only when `AuthPolicy::Mounted { auth_state: Some(_) }`.
  - Full OAuth router mounted in OAuth mode: discovery, JWKS, authorize/callback, token, `/register` (dynamic client registration), and `/auth/login`. Authorize accepts space‑separated multi‑scope values.
  - Fail‑closed scope checks on MCP `tools/call` and `tools/list` via `AuthContext` (`syslog:read` for search/tail/errors/hosts/correlate/stats; `help` is open; `syslog:admin` implies read). Loopback dev bypasses checks; missing auth extensions are logged; failure messages include the action.
  - `PUBLIC_URL` host/origin automatically added to RMCP allowed hosts and CORS. Standard ports (https:443, http:80) include both bare host and host:port variants. `/health` stays unauthenticated.
  - Stdio mode: query‑only forces `LoopbackDev`; non‑loopback safety gate is skipped for stdio. If `AUTH_MODE=oauth` is set in stdio, we warn that it’s ignored.
  - Scripts: `backup.sh` also saves `auth.db` and `auth-jwt.pem` with 0600 perms and tolerates SQLITE_BUSY during WAL checkpoint; `reset-db.sh` adds `--include-auth` and fixes usage/completion edge cases.
  - Tests/docs: added integration and JWT tests (expired/wrong‑iss/wrong‑key/wrong‑aud), discovery body validation, and `/.well-known/oauth-protected-resource` coverage. OAuth setup docs and an rmcp extension propagation regression test are included.
  - Dependencies: bumped `rusqlite` to `0.39` and `r2d2_sqlite` to `0.33`; `lab-auth` is pinned via `git+rev` (updated to the latest merged main).
  - Observability merge: `AppState` and `RuntimeCore` carry `observability` from main; routers updated to pass it through (no behavior change). Fixed missing `.await` on stdio query‑only load in `main.rs`.

- **Migration**
  - To enable OAuth: set `[mcp.auth.mode = "oauth"]`, provide `PUBLIC_URL`, `GOOGLE_CLIENT_ID`, `GOOGLE_CLIENT_SECRET`, and configure `admin_email` or `allowed_emails`; startup fails if incomplete. Dynamic client registration is available at `/register`.
  - Stdio mode needs no credentials (forced `LoopbackDev`). Bearer‑only continues to work; the token is enforced by the shared `AuthLayer` and MCP scope checks apply.
  - If binding to non‑loopback hosts, enable auth (OAuth or bearer token) or startup will fail.
  - Set `SYSLOG_MCP_PUBLIC_URL` so allowed hosts/CORS are derived correctly.
  - Persist and back up `auth.db` and `auth-jwt.pem` (0600 perms). Assign OAuth users the needed scopes (`syslog:read`, `syslog:admin`).

<sup>Written for commit 3f5ded3d384f6dd18cd71faefffe5ab665a56d89. Summary will update on new commits.</sup>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,128 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.17.0] - 2026-05-07
```
[the remark this answers is not in the document]
Comment (jmagar) [reply CHANGELOG.md:10]: Fixed in 1d052d1

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,96 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  - `tools/list` requires `AuthContext` when Mounted but no scope (MCP spec conformance: clients
+    must be able to discover the tool before authenticating to call it, but only if they hold a
+    valid credential).
+  - Denied invocations logged at `warn` level with `subject` + `action` for audit trail.
+  - Pattern: `AuthContext` read from `ctx.extensions.get::<axum::http::request::Parts>()?.extensions.get::<AuthContext>()`
+    (Pattern (a) locked by spike syslog-mcp-brt0.10; no AppState map, no task-local needed).
+  - 9 new unit tests covering all branches: LoopbackDev permit, Mounted+read scope, Mounted+admin
+    scope alone (denied for reads), Mounted+both scopes, empty scopes+read (denied), empty
+    scopes+help (permitted), missing AuthContext fail-closed, tools/list with AuthContext, and
+    scope-check-before-DB verification.
```
[the remark this answers is not in the document]
Comment (jmagar) [reply CHANGELOG.md]: Fixed in 1d052d1

File: CLAUDE.md [hunk 3]
```diff
@@ -106,6 +106,13 @@ SYSLOG_MCP_RECOVERY_FREE_DISK_MB=768  # cleanup target after free-disk breach
 SYSLOG_MCP_CLEANUP_INTERVAL_SECS=60   # storage-budget enforcement interval (>= 5)
 SYSLOG_MCP_CLEANUP_CHUNK_SIZE=1000    # rows deleted per enforcement cycle
 
+# OAuth / JWT auth (disabled by default — set AUTH_MODE=oauth to activate)
```
[the remark this answers is not in the document]
Comment (jmagar) [reply CLAUDE.md]: Fixed in 1d052d1

File: config.toml [hunk 4]
```diff
@@ -27,6 +27,50 @@ server_name = "syslog-mcp"
...
+# ----------------------------------------------------------------------------
+# OAuth / JWT authentication policy.
+#
+# Env vars:
+#   SYSLOG_MCP_AUTH_MODE              bearer (default) | oauth
+#   SYSLOG_MCP_PUBLIC_URL             required when mode=oauth
+#   SYSLOG_MCP_GOOGLE_CLIENT_ID       required when mode=oauth
+#   SYSLOG_MCP_GOOGLE_CLIENT_SECRET   required when mode=oauth
+#   SYSLOG_MCP_TOKEN                  static break-glass bearer (existing)
+#
```
[the remark this answers is not in the document]
Comment (jmagar) [reply config.toml:45]: Fixed in 1d052d1

File: scripts/backup.sh [hunk 5]
```diff
@@ -17,6 +20,12 @@ BACKUP_DIR="${1:-./backups}"
 TIMESTAMP=$(date -u +%Y-%m-%d-%H%M%S)
 BACKUP_FILE="${BACKUP_DIR}/syslog-${TIMESTAMP}.db"
 
+# Auth artifacts live next to the syslog DB by default. Override with the same
+# env vars the syslog-mcp binary honors so the script tracks runtime config.
+DB_DIR="$(dirname "$DB_PATH")"
+AUTH_DB_PATH="${SYSLOG_MCP_AUTH_SQLITE_PATH:-${DB_DIR}/auth.db}"
+AUTH_KEY_PATH="${SYSLOG_MCP_AUTH_KEY_PATH:-${DB_DIR}/auth-jwt.pem}"
+
```
[the remark this answers is not in the document]
Comment (jmagar) [reply scripts/backup.sh]: Fixed in 1d052d1

File: scripts/backup.sh [hunk 6]
```diff
@@ -33,7 +42,38 @@ sqlite3 "$DB_PATH" ".backup '${ESCAPED_BACKUP_FILE}'"
...
+# auth.db is a separate SQLite store managed by lab-auth. Use the same
+# `.backup` (online, WAL-safe) approach so a running syslog-mcp doesn't need
+# to be stopped to capture a consistent OAuth state snapshot.
+if [[ -f "$AUTH_DB_PATH" ]]; then
+    AUTH_BACKUP_FILE="${BACKUP_DIR}/auth-${TIMESTAMP}.db"
+    ESCAPED_AUTH_BACKUP="${AUTH_BACKUP_FILE//\'/\'\'}"
+    # Explicit checkpoint first so even a sidecar copy of the WAL would be
+    # consistent. The .backup command itself is WAL-safe but the checkpoint is
+    # cheap insurance for environments where operators sidestep this script.
+    sqlite3 "$AUTH_DB_PATH" "PRAGMA wal_checkpoint(FULL);" >/dev/null
```
[the remark this answers is not in the document]
Comment (jmagar) [reply scripts/backup.sh]: Fixed in 1d052d1

File: scripts/reset-db.sh [hunk 7]
```diff
@@ -104,4 +137,29 @@ rm -f "$DB_PATH" "${DB_PATH}-wal" "${DB_PATH}-shm"
...
+    if [[ -f "$AUTH_KEY_PATH" ]]; then
+        AUTH_KEY_BACKUP="${BACKUP_DIR}/auth-jwt-pre-reset-${TIMESTAMP}.pem"
+        cp -p "$AUTH_KEY_PATH" "$AUTH_KEY_BACKUP"
+        chmod 600 "$AUTH_KEY_BACKUP" 2>/dev/null || true
+        rm -f "$AUTH_KEY_PATH"
+        echo "Auth JWT key backup + reset complete: ${AUTH_KEY_BACKUP}"
+    else
+        echo "Auth JWT key not found at ${AUTH_KEY_PATH}; nothing to reset"
+    fi
+    echo "Auth state reset: every issued access/refresh token is now invalid."
```
[the remark this answers is not in the document]
Comment (jmagar) [reply scripts/reset-db.sh]: Fixed in 1d052d1

File: scripts/reset-db.sh [hunk 8]
```diff
@@ -5,37 +5,51 @@
...
   --backup-dir DIR  Directory for the timestamped backup file (default: ./backups)
   --force           Skip the interactive confirmation prompt
+  --include-auth    Also wipe lab-auth state (auth.db + auth-jwt.pem)
   --help            Show this help text
 
 Environment:
-  SYSLOG_MCP_DB_PATH   SQLite database path (default: ./data/syslog.db)
+  SYSLOG_MCP_DB_PATH            SQLite syslog database path (default: ./data/syslog.db)
+  SYSLOG_MCP_AUTH_SQLITE_PATH   lab-auth SQLite path (default: <db dir>/auth.db)
+  SYSLOG_MCP_AUTH_KEY_PATH      lab-auth JWT key path (default: <db dir>/auth-jwt.pem)
```
[the remark this answers is not in the document]
Comment (jmagar) [reply scripts/reset-db.sh:51]: Fixed in 1d052d1

File: src/config_tests.rs [hunk 9]
```diff
@@ -484,3 +526,327 @@ fn docker_ingest_rejects_insecure_http_without_explicit_opt_in() {
...
+    cfg.mcp.auth.mode = AuthMode::OAuth;
+    cfg.mcp.auth.public_url = Some("https://syslog.example.com".into());
+    cfg.mcp.auth.google_client_id = Some("id".into());
+    cfg.mcp.auth.google_client_secret = Some("secret".into());
+    cfg.mcp.auth.allowed_emails = vec!["admin@example.com".into()];
+    validate_auth_config(&cfg, true).expect("oauth unlocks non-loopback bind");
+}
+
+#[test]
+fn auth_mode_parses_lowercase_only() {
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/config_tests.rs:885]: Fixed in 1d052d1

File: src/mcp/rmcp_server.rs [hunk 10]
```diff
@@ -265,6 +378,48 @@ fn push_host_variants(hosts: &mut Vec<String>, host: &str, port: u16) {
...
+    let explicit_port = parsed.port();
+    push_host_variants(hosts, host, explicit_port.unwrap_or(listen_port));
+    // Also add with the URL's explicit port if it's non-standard, so both
+    // `host` and `host:port` appear in the list for reverse-proxy deployments
+    // where the Host header may or may not include the port.
+    if let Some(p) = explicit_port {
+        let with_port = format!("{host}:{p}");
+        if !hosts.contains(&with_port) {
+            hosts.push(with_port);
+        }
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/mcp/rmcp_server.rs:440]: Fixed in 1d052d1

File: src/mcp/rmcp_server_tests.rs [hunk 11]
```diff
@@ -278,3 +343,363 @@ async fn rmcp_correlate_events_preserves_truncation_and_host_grouping() {
...
+        auth: crate::config::AuthConfig {
+            public_url: Some("https://syslog.example.com:8443".into()),
+            ..Default::default()
+        },
+    };
+
+    let hosts = allowed_hosts(&config);
+    assert!(
+        hosts.contains(&"syslog.example.com".to_string())
+            || hosts.contains(&"syslog.example.com:8443".to_string()),
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/mcp/rmcp_server_tests.rs]: Fixed in 1d052d1

File: src/mcp/routes.rs [hunk 12]
```diff
@@ -28,62 +27,98 @@ pub fn router(state: AppState) -> Router {
...
+    // that axum's type-checker sees a consistent Router<AppState> merge target.
+    //
+    // Locked Decision: OAuth router only when auth_state: Some(_).
+    // bearer-only (auth_state: None) and LoopbackDev have no OAuth routes.
+    //
+    // Locked Decision: /register and /auth/login are NOT in bearer_only_router
+    // (confirmed by lab-auth's BEARER_ONLY_ROUTER_FORBIDDEN_PATHS snapshot test).
+    let oauth_router: Option<Router> = if let AuthPolicy::Mounted {
+        auth_state: Some(ref state_arc),
+    } = state.auth_policy
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/mcp/routes.rs:66]: Fixed in 1d052d1

File: src/mcp/routes.rs [hunk 13]
```diff
@@ -28,62 +27,98 @@ pub fn router(state: AppState) -> Router {
...
+    let health_state = state.clone();
+    let health_route = Router::new().route("/health", get(health));
 
-    Router::new()
+    let base_with_state: Router<()> = Router::new()
         .merge(authenticated)
-        .merge(unauthenticated)
+        .merge(health_route)
+        .with_state(health_state);
+
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/mcp/routes.rs:105]: Fixed in 1d052d1

File: src/runtime.rs [hunk 14]
```diff
@@ -318,6 +334,231 @@ impl RuntimeCore {
...
+    }
+    if let Some(id) = auth.google_client_id.as_deref() {
+        push_var(&mut vars, "SYSLOG_MCP_GOOGLE_CLIENT_ID", id);
+    }
+    if let Some(secret) = auth.google_client_secret.as_deref() {
+        push_var(&mut vars, "SYSLOG_MCP_GOOGLE_CLIENT_SECRET", secret);
+    }
+    if !auth.admin_email.is_empty() {
+        push_var(&mut vars, "SYSLOG_MCP_AUTH_ADMIN_EMAIL", &auth.admin_email);
+    }
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/runtime.rs:461]: Fixed in 1d052d1

File: src/mcp/rmcp_server.rs [hunk 15]
```diff
@@ -89,9 +109,9 @@ impl ServerHandler for SyslogRmcpServer {
                     error_class = "tool_execution",
                     "MCP tool execution failed"
                 );
-                Ok(CallToolResult::error(vec![Content::text(
-                    "Tool execution failed",
-                )]))
+                Ok(CallToolResult::error(vec![Content::text(format!(
+                    "Tool execution failed for action '{tool_name}'. Check server logs for details."
```
[the remark this answers is not in the document]
Comment (jmagar) [reply src/mcp/rmcp_server.rs]: Fixed in c2b030f
